### ISSUE TITLE
IndexOutOfRangeException while saving dicom file when Specific Character Set (0008,0005) is empty

### ISSUE THREAD

**Issue by gunjitchhhatwal** [fo-dicom/fo-dicom#1879]
**Describe the bug**
An exception is thrown when FellowOakDicom.DicomFile.SaveAsync is called for a dicom dataset having "Value 1 of the Attribute Specific Character Set (0008,0005)" = an empty string. 

System.IndexOutOfRangeException 

   at FellowOakDicom.DicomEncoding.EncodeString(String value, Encoding[] encodings, Boolean isPersonName)
   at FellowOakDicom.DicomStringElement.StringToBytes()
   at FellowOakDicom.IO.Buffer.LazyByteBuffer.get_Size()
   at FellowOakDicom.IO.Writer.DicomWriter.OnElementAsync(DicomElement element)
   at FellowOakDicom.DicomDatasetWalker.DoWalkAsync(IDicomDatasetWalker walker, Queue`1 items)
   at FellowOakDicom.IO.Writer.DicomFileWriter.WriteDatasetAsync(IByteTarget target, DicomTransferSyntax syntax, DicomDataset dataset, DicomWriteOptions options)
   at FellowOakDicom.IO.Writer.DicomFileWriter.WriteAsync(IByteTarget target, DicomFileMetaInformation fileMetaInfo, DicomDataset dataset)
   at **FellowOakDicom.DicomFile.SaveAsync**(Stream stream, DicomWriteOptions options)

**To Reproduce**

```
public async static Task TestDicom()
{
    var dicom = DicomFile.Open("test.dcm"); // any valid dcm file with an attribute having a string value (e.g. VR = SH or LO). E.g. Accession Number

    dicom.Dataset.Add(DicomTag.SpecificCharacterSet, ""); // add (0008,0005) tag with empty string as value

    Stream stream = new MemoryStream();

    await dicom.SaveAsync(stream);// try to save the file to a stream, throws exception here
}
```

**Expected behavior**
Per DICOM standard, an empty value 1 of this attribute essentially represents the default character repertoire. One should be able to save the file to the stream as if the specific character set tag was not present at all. 

https://dicom.nema.org/medical/dicom/current/output/chtml/part05/chapter_6.html 

"_The replacement Character Repertoire specified in Value 1 of the Attribute Specific Character Set (0008,0005) (**or the Default Character Repertoire if Value 1 is empty**) may be further extended with additional Coded Character Sets_"

**Screenshots or test DICOM files**
![image](https://github.com/user-attachments/assets/2f18f862-a440-4db6-be5e-e9012eed29fc)

**Environment**
Fellow Oak DICOM version: 5.1.4
OS: (e.g. Windows 11 x64)
Platform: .NET 6.0

**Comment by mrbean-bremen** [comment]
Looks like I have introduced this bug... I will have a look later.

**Comment by gunjitchhhatwal** [comment]
Confirming that it is not throwing this exception in v5.1.3 for the same scenario.

**Comment by gofal** [comment]
I also found this bug yesterday. here, the Specific Character Set, that has an empty value, is located within an Annotation Sequence:
![image](https://github.com/user-attachments/assets/b4e049b7-956d-4891-96f9-20dca7ef1de1)

I wonder, if this value is empty, should the the default (meaning ASCII) be taken, or should the encoding of the parent Dataset be taken?
So what I am asking is, waht the "default" means that should be taken if the value is empty. Is really the DICOM default Encoding meant, or does it mean the encoding that we would take as if the tag was not present at all?

**Comment by gofal** [comment]
because then, the fix would rather be in DicomDataset.GetEncodingForSerialization()

    internal Encoding[] GetEncodingsForSerialization()
    {
      return TryGetValues<string>(DicomTag.SpecificCharacterSet, out var charsets) **&& charsets.Any()**
          ? DicomEncoding.GetEncodings(charsets)
          : FallbackEncodings;
    }

where the value of the SpecificCharacterset is only taken if there is a value present, and otherwise the FallbackEncoding of the DicomDataset is chosen, which is either 
- the FallbackEncoding that the user passed as Parameter to the DicomDataset
- the encoding of the parent Dataset if the current dataset is contained in a Sequence
- or else the DICOM-default-encoding, namely ASCII

**Comment by mrbean-bremen** [comment]
> I wonder, if this value is empty, should the the default (meaning ASCII) be taken, or should the encoding of the parent Dataset be taken?

As far as I understand, the parent encoding is always overwritten if it is set in the sequence, even if it is set to an empty value (which means ASCII encoding). As far as I understand the code, this is what it already does. 

Note the fix (and this issue) is about writing to a dataset, not reading it. Reading a dataset with an empty Specific Character Set should work fine.

**Comment by gofal** [comment]
Thanks, I was unsure about that and misunderstood the Standard. When I read it, then I understood that the Empty Value 1 is only supposed to be treated as the Ansi-Encoding in terms of Code Extension Techniques. Just as my favorite dicom-lookup-website https://dicom.innolitics.com/ciods/ct-image/sop-common/00080005 does: "Note
If the Attribute Specific Character Set (0008,0005) has more than one value and value 1 is empty, it is assumed that value 1 is ISO 2022 IR 6."

But when reading the Dicom Standard then I was not able to find the requirement, that this is only in conjunction with Code Extension. 

Sure, when opening a file from drive or receiving a dataset via network, then all values are kept in memory as binary stream. Only when trying to write or also when accessing the content of a DicomTag, then the binary data has to be encoded or decoded. and therefore the current encoding is evaluated and this was then where the exception happened.

**Comment by gofal** [comment]
Thanks, your PR fixes the issue. when it comes to actual encoding and no encoding (for whatever reason) is passed as parameter, then the default-encoding is used. this also might fix other scenarios where encoding of items may be used in other contextes.

But because here the DICOM has a clear claim: when the 0008,0005 is present and empty, then DefaultEncoding has to be used.
So would you mind, to make this clear in Code? 
It's the method `DicomEncoding.GetEncodings(string[] charsets)` that is called when a DicomDataset or an Observer actually has the content of 0008,005 and wants to know, which encoding to use.
Currently this method returns `[]` (empty array), which is then propagated through the application and then later causes the IndexOutOfRangeException.
Basically, this method should reference the DICOM-Standard and in case the `charsets` array is empty, it should return the DefaultEncodings, shouldn't it?

Both changes, the suggested change in method GetEncodings(charsets) and then the fallback in `GetStrictEncodings` would fix the issue very clearly, wouldn't it?

**Comment by mrbean-bremen** [comment]
Thanks, good point - you are probably right about `GetEncodings` - it should behave the same way as `GetStrictEncodings` does in the PR, e.g. return `DefaultArray` instead of an empty array. I will change this accordingly.

And yes, the DICOM standard is not too explicit about that. It only defines what shall happen when no `Specific Character Set` is present, or if it is present and has at least one value, so you have to infer the behavior.

**Comment by mrbean-bremen** [comment]
Ok, I have adapted `GetEncodings` accordingly and added a couple of tests that had failed before. Note that I did not change the fallback handling for writing, as I think it is correct as is.

**Comment by jaime-olivares** [comment]
This patch is urgent, as it is breaking the existing functionality that was OK with 5.1.3
